doc/reference/trd-storage-permissions.md

```diff
@@ -0,0 +1,174 @@
+Application Persistent Data Storage Permissions
+===============================================
+
+**TRD:** <br/>
+**Working Group:** Kernel<br/>
+**Type:** Documentary<br/>
+**Status:** Draft <br/>
+**Author:** Brad Campbell<br/>
+**Draft-Created:** 2024/06/06<br/>
+**Draft-Modified:** 2024/06/06<br/>
+**Draft-Version:** 1<br/>
+**Draft-Discuss:** devel@lists.tockos.org<br/>
+
+
+Abstract
+-------------------------------
+
+Tock supports storing persistent state for applications, and all persistent
+state in Tock is identified based on the application that stored it. Tock
+supports permissions for persistent state, allowing for the kernel to restrict
+which applications can store state and which applications can read stored state.
+This TRD describes the permissions architecture for persistent state in Tock.
+This document is in full compliance with [TRD1][TRD1].
+
+
+1 Introduction
+-------------------------------
+
+Tock applications need to be able to store persistent state. Additionally,
+applications need to be able to keep data private from other applications. The
+kernel should also be able to allow specific applications to read and modify
+state from other applications.
+
+This requires a method for assigning applications persistent identifiers,
+a mechanism for granting storage permissions to specific applications,
+and kernel abstractions for implementing storage capsules that respect
+the storage permissions.
+
+
+2 Scope
+-------------------------------
+
+This document only describes the permission architecture in Tock for supporting
+application persistent storage. This document does not include specific types of
+persistent storage (e.g., flash, FRAM, etc.) or storage access abstractions
+(e.g., block-access, byte-access, etc.).
+
+
+3 Permissions
+-------------------------------
+
+All persistent application data is labeled based on the application which wrote
+the data. By default, applications can always read and modify data they wrote.
+Applications can read and modify data from other applications with suitable
+permissions.
+
+There are three types of permissions:
+
+1. **Write**: The application can write data.
+1. **Read**: The application can read other applications' data.
+1. **Modify**: The application can modify other applications' data.
+
+Each permission type is independent. For example, an application can be given
+read permission for specific data but not be able to write new data itself.
+
+
+4 Requirements
+-------------------------------
+
+The Tock storage model imposes the following requirements:
+
+1. Applications are given separate write, read, and modify permissions.
+2. The label stored with the persistent data when the data are written is the
+   application's short AppID.
```

Comment:
This should use the `writeID`. That is why it exists and how everything has been designed.

A `writeID` gives us more flexibility in terms of say A/B updating applications or other more complex setups

Reply:
I disagree. We need to unify write id and app id, and having two adds confusion.

The original design was added as an experimental feature, and appid was designed with exactly this case in mind (it just didn't exist yet so writeid was needed).

Reply:
Making sure I understand correctly, this replaces what `writeID` currently allows for by  saying that now:
 - We might have apps `A, B, C`
 - We could give `A {read/append}`, `B {read/append/modify}`, and `C {read}`
 - We could give all three apps access to the same storage
 - A record could be created by `A` or `B`, and could only be modified by `B`
 - Any of `A`, `B`, or `C` could determine who last wrote a record because on a read operation the `shortId` which last wrote the record will always be returned [and e.g., if `A` created a record, but `B` modified it, when `A` reads it again it will see that the record was written by `B`]

Reply:
> * Any of `A`, `B`, or `C` could determine who last wrote a record because on a read operation the `shortId` which last wrote the record will always be returned [and e.g., if `A` created a record, but `B` modified it, when `A` reads it again it will see that the record was written by `B`]

I don't know if this is true.

Perhaps we should be explicit about what modify does with the storage id. I think I was thinking it does not change the stored id.

Reply:
I guess the question maybe is the purpose of the ID.

Is the purpose to identify what app "owns" a storage region (implicitly saying that multiple apps can't share/co-own a region)?

Or, as in Alistair's use case, is it so that apps which do share some space can identify who last wrote the record?

Or, are both useful, and there should be two IDs (per record? high overhead? ...per region?)?

Reply:
> Is the purpose to identify what app "owns" a storage region (implicitly saying that multiple apps can't share/co-own a region)?

This one. But they can share, that is why app B can have read and modify access to ShortId(A).

Reply:
> Is the purpose to identify what app "owns" a storage region (implicitly saying that multiple apps can't share/co-own a region)?

The `writeID` is to specify what ID new data is written with. It's hopefully clearly documented here: https://book.tockos.org/doc/tock_binary_format#7-storage-permissions

Apps can share data by using the same `writeID`

Reply:
I'm not clear on why apps shouldn't be able to co-own a region? (As in why have that enforcement)

Reply:
What use case does having two apps create new stored state with the same identifier support that having two apps create new stored state with different identifiers and read & modify permissions for each other's identifier not?

Conversely, using AppId allows us to enforce the Tock threat model since we have a mechanism for cryptographically assigning storage identifiers to applications.

Reply:
I think the proposal is to basically get rid of `write_id`, and rely on App IDs ubiquitously. I think there are several advantages to doing this:

1. It's fewer things to learn/reason/remember.
2. The expected "default" case (an app has private storage), requires fewer moving parts
3. It makes any sharing beyond the default case more explicit---with the semantics of write_ids, any app that comes along in the future that has the same write_id can write to that data, so I don't really have a record anywhere of who is allowed to write to the data (it's emergent from the TBF headers of all possible apps that could be installed). Conversely, with explicit sharing (listing AppIDs that can write), the _data_ has an exact record of "who" can access it in one place, relying _only_ on AppIDs uniquely naming "apps".

Reply:
Importantly, I believe that any policy you can implement with `write_id` et al, can also be implemented with App IDs and the flexible abstract permissions trait in this TRD.

Reply:
> What use case does having two apps create new stored state with the same identifier support that having two apps create new stored state with different identifiers and read & modify permissions for each other's identifier not?

It's not about creating new stored state. It's update updating an existing application.

If you have an application already running and that has already written data and you want to replace that application via an A/B update it is a lot easier if you can use the same write_id.

Basically it allows more flexibility of adding functionality in the future

Reply:
> If you have an application already running and that has already written data and you want to replace that application via an A/B update it is a lot easier if you can use the same write_id.

Is the update for the same application or replacing one application with a different application? If the former (same application), then the AppId will remain the same. If the latter (different application), then it would need correct modify permissions.

I agree that in the different app case it is a little more convenient to be able to explicitly set the same write id, although I don't think it is any more flexible (both approachs support this use case). However, replacing an app with a different app and wanting the same storage seems less common than replacing an app with a new version (which would have the same AppId).

Reply:
A/B updating. So a new application that replaces the original one. The idea is to update a live system with no downtime (just a reboot at some point in the future).

If using `shortIDs` don't you end up with data that isn't "owned" by any application? I guess app B is allowed to write to it, so it is still accessible, but it feels a bit clunky

Reply:
> If using `shortIDs` don't you end up with data that isn't "owned" by any application?

Yes, but isn't that reasonable? If I replace Firefox with Chrome, it makes sense to me that whatever state Firefox stored is no longer used and isn't given to Chrome (but would be if I re-installed Firefox in the future). If I update Firefox 126 with Firefox 127, then it maintains access to any of its prior stored state (and would with ShortId).

Reply:
Firefox and Chrome are two different applications, not an update. Plus that's a desktop system, not an embedded system.

Maybe a more concrete example will help. Say you have a FIDOv1 application. It works great and everyone is using it. It uses writeID and readID `A`.

You want to update to FIDOv2 support. To avoid breaking anything, FIDOv2 is a new application. So you flash FIDOv2 application with writeID and readID `A`. Now two applications are running.

For legacy FIDOv1 operations the FIDOv1 application handles it. For FIDOv2 ops the new FIDOv2 application handles it. But the data is stored in the same place/region/store (basically they share the data, the FIDOv1 application just never parses the new data as it doesn't understand it).

Eventually the FIDOv1 application is removed and the data is all migrated to the FIDOv2 app.

The key part is that the FIDOv1 application is never updated. Let's say due to compliance, code or testing reasons.

That above works fine with with this TRD.

The tricky part is what happens if you incrementally want to support FIDOv1 in the new FIDOv2 application. Or if you only have partial support. So you maybe enrol new FIDOv1 requests with the new application, but handle the challenges in the legacy one. You basically want to fallback to the legacy FIDOv1 application in some situations. Data can be written by the new application with the original writeID (`A`) to maintain compatibility.

Using a fixed ShortID, which has to be unique, backs you into a corner here.

I agree it isn't super common, and most use cases don't care. But for a small select few it matters. Especially if you are unable or unwilling to update existing applications. Which in embedded, certified systems can be hard

Reply:
If I'm understanding correctly the use case is:

- A new application (FIDOv2) needs to be able to write new data that a legacy application (FIDOv1) can access. It may not be possible or easy to update the storage permissions for the legacy application (FIDOv1).

Sharing a WriteID makes this trivial because the two separate applications effectively share all state.

I'm concerned about complicating the entire storage design in Tock because of this use case. But:

In Section 8 I outline a few examples of what different storage systems might exist. It seems to me there are generally going to be three classes:

1. State which is stored and retrieved based on some record-specific name. For example, using a filename, a key name (in a key-value pair), or a record name.
2. State which is stored and retrieved based on its use case or based on which application "owns" it (ie wrote it originally, or uses it exclusively). For example, configuration state, arbitrary per-application state (like appstate.rs), openthread's dataset, or LoRa connection parameters/keys (also essentially per-application state).
3. State which is in a shared log. For example, logging. (I'm not sure if there are more examples of this.)

I can imagine implementing the underlying storage for the FIDOv1/2 example with either options 1 or 2.

With a storage system of type 2, each app (FIDOv1 and FIDOv2) would have their own storage region they can use. They each access it the stored state in its entirety, and probably have some data structure in the stored state. In the upgrade scenario, to allow FIDOv2 to save new requests that the FIDOv1 app can use, FIDOv2 is installed with modify permissions for `ShortId(FIDOv1)`. Since the FIDOv1 state already exists, FIDOv2 just needs to modify it.

Using a storage system of type 1 seems like where the issue of not sharing writeIDs becomes a challenge. Any new fido request handled by the new app must be saved as a new object with a unique name (hence it would have FIDOv2's ShortId as its storage identifier). But, even if FIDOv2 _could_ write with FIDOv1's writeID, it would need to store with a name such that FIDOv1 would find it and be able to use it.

I'm trying to judge how likely it is that this type of backwards compatibility use case would use named storage (type 1) rather than per-app storage (type 2). It seems to me that it would be hard to implement this with type 1. 

Type 2, on the other hand, seems like a natural fit, and would be easy to support with the permissions model. For example, imagine the storage provides each app with its own table in a database. FIDOv1 stores requests as rows in its table. FIDOv2, with modify access for `ShortId(FIDOv1)` can append rows to that table. FIDOv1 then will get the new data with its normal queries.

My question is, then, how important is supporting this use case on a type 1 storage, where the new app would be required to write completely new objects to the storage system? Is it reasonable to say: to enable this use case either use a type 2 storage or design the kernel to be able to modify storage permissions for legacy apps (ie giving FIDOv1 read access to `ShortId(FIDOv2)`)?

Reply:
> Using a storage system of type 1 seems like where the issue of not sharing writeIDs becomes a challenge. Any new fido request handled by the new app must be saved as a new object with a unique name (hence it would have FIDOv2's ShortId as its storage identifier).

Exactly. And FIDO data would probably be stored in a key-value store (type 1 in your example).

The other point to remember is that when the FIDOv1 app was written, it wasn't written knowing this update would happen. So you can't plan ahead

> But, even if FIDOv2 _could_ write with FIDOv1's writeID, it would need to store with a name such that FIDOv1 would find it and be able to use it.

Yeah, but that would be pretty easy. The keys would just be a hash of the token used in the CTAP data. Both v1 an v2 could easily use the same scheme for the "legacy" FIDOv1 setup.

> I'm trying to judge how likely it is that this type of backwards compatibility use case would use named storage (type 1) rather than per-app storage (type 2). It seems to me that it would be hard to implement this with type 1.

Which is what I'm worried about. I would expect FIDO data to be using named storage (Tock has a key-value store for this use case).

And this is just one example I could easily think of, I suspect there might be a few more complex cases we lock out using shortIDs for the write ID. That might be a trade off worth making, but we should at least list the cases we don't support and justify why we don't support them.

> My question is, then, how important is supporting this use case on a type 1 storage, where the new app would be required to write completely new objects to the storage system? Is it reasonable to say: to enable this use case either use a type 2 storage or design the kernel to be able to modify storage permissions for legacy apps (ie giving FIDOv1 read access to `ShortId(FIDOv2)`)?

I personally don't think so. But it's not the worst thing in the world to say that. **But** I do think that should be written down and documented as something we don't support, with justification on why we chose not to support it

Reply:
> Type 2, on the other hand, seems like a natural fit, and would be easy to support with the permissions model. For example, imagine the storage provides each app with its own table in a database. FIDOv1 stores requests as rows in its table. FIDOv2, with modify access for `ShortId(FIDOv1)` can append rows to that table. FIDOv1 then will get the new data with its normal queries.

Wouldn't the new data be written with the new FIDOv2 write/short ID? Adding something to a database feels like a write (new data) then an append (modify existing data).

I guess you could debate either way, maybe something else we should clarify?